pkg/router/controller/route_secret_manager.go

```diff
@@ -118,106 +192,199 @@ func (p *RouteSecretManager) HandleRoute(eventType watch.EventType, route *route
+			// and re-evaluatation by all the plugins (including this plugin).
+			p.recorder.RecordRouteUpdate(route, "ExternalCertificateSecretUpdated", msg)
 		},
 
 		DeleteFunc: func(obj interface{}) {
 			secret := obj.(*kapi.Secret)
-			msg := fmt.Sprintf("secret %s deleted for route %s/%s", secret.Name, route.Namespace, route.Name)
+			key := generateKey(namespace, routeName)
+			msg := fmt.Sprintf("secret %q deleted for route %q", secret.Name, key)
 			log.V(4).Info(msg)
 
-			// unregister associated secret monitor
-			if err := p.secretManager.UnregisterRoute(route.Namespace, route.Name); err != nil {
-				log.Error(err, "failed to unregister route")
+			// keep the secret monitor active and mark the secret as deleted for this route.
+			p.deletedSecrets.Store(key, true)
+
+			// Ensure fetching the updated route
+			route, err := p.getUpdatedRoute(namespace, routeName)
+			if err != nil {
+				return
 			}
 
+			// Reject this route
 			p.recorder.RecordRouteRejection(route, "ExternalCertificateSecretDeleted", msg)
+			// Stop serving this route
 			p.plugin.HandleRoute(watch.Deleted, route)
 		},
 	}
 }
 
+// validate checks the route's external certificate configuration.
+// If the validation fails, it records the route rejection and triggers
+// the deletion of the route by calling the HandleRoute method with a watch.Deleted event.
+//
+// NOTE: TLS data validation and sanitization are handled by the next plugin `ExtendedValidator`,
+// by reading the "tls.crt" and "tls.key" added by populateRouteTLSFromSecret.
+func (p *RouteSecretManager) validate(route *routev1.Route) error {
+	fldPath := field.NewPath("spec").Child("tls").Child("externalCertificate")
+	if err := routeapihelpers.ValidateTLSExternalCertificate(route, fldPath, p.sarClient, p.secretsGetter).ToAggregate(); err != nil {
+		log.Error(err, "skipping route due to invalid externalCertificate configuration", "namespace", route.Namespace, "route", route.Name)
+		p.recorder.RecordRouteRejection(route, "ExternalCertificateValidationFailed", err.Error())
+		p.plugin.HandleRoute(watch.Deleted, route)
+		return err
+	}
+	return nil
+}
+
+// populateRouteTLSFromSecret updates the TLS configuration of the route using data from the referenced secret.
+// If fetching the secret fails, it records the route rejection and triggers
+// the deletion of the route by calling the HandleRoute method with a watch.Deleted event.
+// Note: This function performs an in-place update of the route. The caller should be aware that the route's TLS configuration will be modified directly.
+func (p *RouteSecretManager) populateRouteTLSFromSecret(route *routev1.Route) error {
+	// read referenced secret
+	secret, err := p.secretManager.GetSecret(context.TODO(), route.Namespace, route.Name)
+	if err != nil {
+		log.Error(err, "failed to get referenced secret")
+		p.recorder.RecordRouteRejection(route, "ExternalCertificateGetFailed", err.Error())
+		p.plugin.HandleRoute(watch.Deleted, route)
+		return err
```

Comment:
Same suggestion:
```suggestion
		return fmt.Errorf("failed to get referenced secret: %w", err)
```

Reply:
Due to the same reason stated above, I would like to keep them inside a function to avoid code redundancy.